bitcoin/src/address.rs

```diff
@@ -469,6 +469,28 @@ impl Payload {
         }
     }
 
+    /// Returns true if the address creates a particular script
+    pub fn matches_script_pubkey(&self, script: &Script) -> bool {
+        match *self {
+            Payload::PubkeyHash(ref hash) => if script.is_p2pkh() {
+                &script.as_bytes()[3..23] == <PubkeyHash as AsRef<[u8; 20]>>::as_ref(hash)
+            } else {
+                false
+            },
+            Payload::ScriptHash(ref hash) => if script.is_p2sh() {
+                 &script.as_bytes()[2..22] == <ScriptHash as AsRef<[u8; 20]>>::as_ref(hash)
+            } else {
+                false
+            }
+            Payload::WitnessProgram(ref prog) => if script.is_witness_program() {
+                &script.as_bytes()[2..] == prog.program.as_bytes()
+            } else {
+                false
+            },
+        }
+
+    }
```

Comment:
Perhaps this is cleaner?

```rust
    pub fn matches_script_pubkey(&self, script: &Script) -> bool {
        match *self {
            Payload::PubkeyHash(ref hash) => if script.is_p2pkh() {
                return &script.as_bytes()[3..23] == <PubkeyHash as AsRef<[u8; 20]>>::as_ref(hash);
            }
            Payload::ScriptHash(ref hash) => if script.is_p2sh() {
                return &script.as_bytes()[2..22] == <ScriptHash as AsRef<[u8; 20]>>::as_ref(hash);
            }
            Payload::WitnessProgram(ref prog) => if script.is_witness_program() {
                return &script.as_bytes()[2..] == prog.program.as_bytes();
            },
        }
        false
    }
```

Reply:
It's shorter for sure, but I personally not a fan of multiple returns. This fn is small though, so why not.

Reply:
As you wish man, no bother either way.

Reply:
Or this:

```rust
    pub fn matches_script_pubkey(&self, script: &Script) -> bool {
        match *self {
            Payload::PubkeyHash(ref hash) if script.is_p2pkh() => {
                &script.as_bytes()[3..23] == <PubkeyHash as AsRef<[u8; 20]>>::as_ref(hash)
            },
            Payload::ScriptHash(ref hash) if script.is_p2sh() => {
                &script.as_bytes()[2..22] == <ScriptHash as AsRef<[u8; 20]>>::as_ref(hash)
            },
            Payload::WitnessProgram(ref prog) if script.is_witness_program() => {
                &script.as_bytes()[2..] == prog.program.as_bytes()
            },
            Payload::PubkeyHash(_) | Payload::ScriptHash(_) | Payload::WitnessProgram(_) => false,
        }
    }
```

Reply:
thank you all for suggestions! updated